src/array.cr

```diff
@@ -2051,14 +2059,16 @@ class Array(T)
     @capacity - @offset_to_buffer
   end
 
-  private def calculate_new_capacity
-    return INITIAL_CAPACITY if @capacity == 0
-
-    if @capacity < CAPACITY_THRESHOLD
-      @capacity * 2
-    else
-      @capacity + (@capacity + 3 * CAPACITY_THRESHOLD) // 4
+  private def calculate_new_capacity(new_size = @capacity + 1)
+    new_capacity = @capacity == 0 ? INITIAL_CAPACITY : @capacity
+    while new_capacity < new_size
+      if new_capacity < CAPACITY_THRESHOLD
+        new_capacity *= 2
+      else
+        new_capacity += (new_capacity + 3 * CAPACITY_THRESHOLD) // 4
+      end
     end
+    new_capacity
```

Comment:
I'm wondering if it makes sense to avoid the loop in this method when possible. I figure it's not necessary when the method is called without arguments, so maybe splitting this method into two overloads could be benefitial? This would mean keep the old implementation and add a new overload which accepts an argument.
Not sure if it makes a difference after LLVM optimizations, though.